USBPcapDriver/USBPcapURB.c

```diff
@@ -673,75 +673,98 @@ VOID USBPcapAnalyzeURB(PIRP pIrp, PURB pUrb, BOOLEAN post,
-                            for (i = 0; i < transfer->NumberOfPackets; i++)
-                            {
-                                /* capture only packet buffers that contain data to minimize capture size */
-                                if ((transfer->IsoPacket[i].Length > 0) &&
-                                    (transferLength < transfer->IsoPacket[i].Offset + transfer->IsoPacket[i].Length))
-                                {
-                                    transferLength = transfer->IsoPacket[i].Offset + transfer->IsoPacket[i].Length;
-                                }
-                            }
-                        }
-                    }
-                    else if (transfer->TransferBufferMDL)
-                    {
-                        /* If the transfer buffer is an MDL, the buffer length can be retrieved directly. */
-                        transferLength = MmGetMdlByteCount(transfer->TransferBufferMDL);
-                    }
+                    compactedLength += transfer->IsoPacket[i].Length;
                 }
             }
-            else
+
+            if (compactedLength > transfer->TransferBufferLength)
             {
-                if (post == FALSE)
-                {
-                    /* Write to device, on its way to controller */
-                    transferLength = (UINT32)transfer->TransferBufferLength;
-                }
-			}
+                /* This is a safety check -- the numbers don't add up (this should never happen) */
+                DkDbgStr("Sum of Isochronous transfer packet lengths exceeds transfer buffer length");
+                ExFreePool((PVOID)packetHeader);
+                break;
+            }
 
-            if (transferLength == 0)
+            if (compactedLength == 0)
             {
+                /* No data was transferred */
                 packetHeader->header.dataLength = 0;
-                transferBuffer = NULL;
+                compactedBuffer = NULL;
+
+                /* Copy the packet headers untouched */
+                for (i = 0; i < transfer->NumberOfPackets; i++)
+                {
+                    packetHeader->packet[i].offset = transfer->IsoPacket[i].Offset;
+                    packetHeader->packet[i].length = transfer->IsoPacket[i].Length;
+                    packetHeader->packet[i].status = transfer->IsoPacket[i].Status;
+                }
             }
             else
             {
-                packetHeader->header.dataLength = transferLength;
+                /* Compact the data to minimize the capture size */
+                packetHeader->header.dataLength = (UINT32)compactedLength;
+
+                /* Allocate a buffer for the compacted results */
+                /* The original data must remain untouched */
+                compactedBuffer = ExAllocatePoolWithTag(NonPagedPool,
+                    (SIZE_T)compactedLength,
+                        'COSI');
+
+                if (compactedBuffer == NULL)
```

Comment:
OOM failure is treated the same as it was when trying to allocate a packet header. When the kernel gets to the point where it can't allocate memory, the best approach may be to just drop the packet and move on.